### PR TITLE
fixing the policy format changes issue

### PR DESCRIPTION
**Pull request by parweza**
Testing Done:
manually created the policy resource and validated 
1. Apply does not update the policy even though format difference
2. server side policy update gets overwritten  on apply
[update.txt](https://github.com/oracle/terraform-provider-oci/files/1342432/update.txt)
[create.txt](https://github.com/oracle/terraform-provider-oci/files/1342433/create.txt)
[terraform.tfstate.backup.txt](https://github.com/oracle/terraform-provider-oci/files/1342442/terraform.tfstate.backup.txt)
[terraform.tfstate.txt](https://github.com/oracle/terraform-provider-oci/files/1342443/terraform.tfstate.txt)
[policy.tf.txt](https://github.com/oracle/terraform-provider-oci/files/1342444/policy.tf.txt)

### CODE REVIEW

File: resource_obmcs_identity_policy.go [hunk 1]
```diff
@@ -165,23 +205,31 @@ func (s *PolicyResourceCrud) Update() (e error) {
...
 	s.D.Set("name", s.Res.Name)
 	s.D.Set("description", s.Res.Description)
 	s.D.Set("compartment_id", s.Res.CompartmentID)
 	s.D.Set("state", s.Res.State)
 	s.D.Set("time_created", s.Res.TimeCreated.String())
 }
 
+func getMD5Hash(values [] string) string {
+	statements := strings.Join(values, "#");
+	return fmt.Sprintf("%x", md5.Sum([]byte(statements)))
```
Comment (codycushing) [inline resource_obmcs_identity_policy.go]: Have a look at `hex.EncodeToString`

Comment (parweza) [reply resource_obmcs_identity_policy.go]: hex.EncodeToString  encodes the string into hexadecimal format, I want hash so that we can do equality comparison without having to storE the entire policy statement.

 Are you suggesting to do hex encoding before hashing or do you want me to store encoded string?

Comment (codycushing) [reply resource_obmcs_identity_policy.go]: You are using fmt.Sprintf to coerce a hex value to a string, I was suggesting using the intended utility for that conversion

**Comment by codycushing** [comment]
What do you think about including this as an acceptance test case?

File: resource_obmcs_identity_policy.go (hunk 1 above)
Comment (parweza) [reply resource_obmcs_identity_policy.go]: Oh, Ok. Done

**Comment by parweza** [comment]
Updated Test cases

File: resource_obmcs_identity_policy_test.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func (s *ResourceIdentityPolicyTestSuite) SetupTest() {
 	s.Config = testProviderConfig() + `
 	resource "oci_identity_compartment" "t" {
 		name = "-tf-compartment"
-		description = "tf test compartment"
+		description = "tf test compartment2"
```
Comment (codycushing) [inline resource_obmcs_identity_policy_test.go]: Is this change required?

File: resource_obmcs_identity_policy_test.go [hunk 3]
```diff
@@ -67,6 +70,25 @@ func (s *ResourceIdentityPolicyTestSuite) TestAccResourceIdentityPolicy_basic()
...
+					statements = ["Allow group ${oci_identity_group.t.name} to read instances in    compartment ${oci_identity_compartment.t.name}"]
+				}`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet(s.ResourceName, "id"),
+					resource.TestCheckResourceAttrSet(s.ResourceName, "compartment_id"),
+					resource.TestCheckResourceAttrSet(s.ResourceName, "time_created"),
+					resource.TestCheckResourceAttr(s.ResourceName, "name", "-tf-policy"),
+					resource.TestCheckResourceAttr(s.ResourceName, "policyHash", "5f251343112c766d4ef6014f0c97c68c"),
+					resource.TestCheckResourceAttr(s.ResourceName, "description", "automated test policy"),
+					resource.TestCheckResourceAttr(s.ResourceName, "statements.#", "1"),
```
Comment (codycushing) [inline resource_obmcs_identity_policy_test.go]: I don't think we should duplicate assertions from previous steps for values that are not being changed, unless there is a reasonable suspicion that some change might have side effects.

File: resource_obmcs_identity_policy_test.go [hunk 4]
```diff
@@ -51,12 +51,15 @@ func (s *ResourceIdentityPolicyTestSuite) TestAccResourceIdentityPolicy_basic()
 					compartment_id = "${oci_identity_compartment.t.id}"
 					name = "-tf-policy"
 					description = "automated test policy"
-					statements = ["Allow group ${oci_identity_group.t.name} to read instances in compartment ${oci_identity_compartment.t.name}"]
+					statements = ["Allow group ${oci_identity_group.t.name} to read instances in    compartment ${oci_identity_compartment.t.name}"]
```
Comment (codycushing) [inline resource_obmcs_identity_policy_test.go]: It might be clearer if this is tested in a dedicated update step, rather than implicitly in the creation step. However, since creation appears to be one of the areas this can fail, it could be clearer to split this into another test function, like TestAccResourceIdentityPolicy_postApply()

File: resource_obmcs_identity_policy_test.go (hunk 2 above)
Comment (parweza) [reply resource_obmcs_identity_policy_test.go]: someone updated the description of compartment, this change was to make the test runnable while I fix the compartment description.
Fixed.

File: resource_obmcs_identity_policy_test.go (hunk 3 above)
Comment (parweza) [reply resource_obmcs_identity_policy_test.go]: removed, kept only assertions that are validates that field values has not changed

File: resource_obmcs_identity_policy_test.go (hunk 4 above)
Comment (parweza) [reply resource_obmcs_identity_policy_test.go]: update the tests as discussed

File: resource_obmcs_identity_policy.go [hunk 5]
```diff
@@ -69,6 +85,27 @@ func PolicyResource() *schema.Resource {
...
+	return func(k string, old string, new string, d *schema.ResourceData) bool {
+		oldHash := getOrDefault(d, "policyHash", "")
+		newHash := getMD5Hash(toStringArray(d.Get("statements")))
+		oldETag := getOrDefault(d, "lastUpdateETag", "")
+		currentETag := getOrDefault(d, "ETag", "")
+		suppressDiff := strings.EqualFold(oldHash, newHash) && strings.EqualFold(oldETag, currentETag)
+
+		return suppressDiff
+	}
+}
```
Comment (codycushing) [inline resource_obmcs_identity_policy.go]: Why not just implement SchemaDiffSuppressFunc with the ignorePolicyFormatDiff function and pass it to the schema directly?

File: resource_obmcs_identity_policy.go [hunk 6]
```diff
@@ -69,6 +85,27 @@ func PolicyResource() *schema.Resource {
...
+
+func getOrDefault(d *schema.ResourceData, key string, defaultValue string) string {
+	valueString := defaultValue
+	if value, ok := d.GetOk(key); ok {
+		valueString = value.(string)
+	}
+
+	return valueString;
```
Comment (codycushing) [inline resource_obmcs_identity_policy.go]: no semicolon

**Comment by codycushing** [comment]
Be sure to run `make fmt` or install the pre-push hook

File: resource_obmcs_identity_policy.go (hunk 5 above)
Comment (parweza) [reply resource_obmcs_identity_policy.go]: done
